Android cmake

freedomtan opened the pull request:
two fixes after cmake rule refactoring
* Add -s for Android back
* fix -DBUILD_TEST=ON -DBUILD_BINARY=ON for Android

freedomtan committed:
Add -s for Android back

Android libraries are statically linked, we'd better strip binaries

freedomtan committed:
fix -DBUILD_TEST=ON -DBUILD_BINARY=ON for Android

make
```
./script/build_android -DBUILD_TEST=ON -DBUILD_BINARY=ON
```
work

Yangqing commented:
@Maratyszcza , what's your take on the -s option?

Yangqing commented on caffe2/CMakeLists.txt:
```diff
@@ -126,11 +126,9 @@ if (BUILD_TEST)
-    target_link_libraries(${test_name} ${Caffe2_MAIN_LIBS})
-    target_link_libraries(${test_name} gtest_main)
-    target_link_libraries(${test_name} ${Caffe2_DEPENDENCY_LIBS})
+    target_link_libraries(${test_name} ${Caffe2_MAIN_LIBS} gtest_main)
```
The left lines are actually all needed - see comments above - could you add them back?

freedomtan replied:
The reason to remove them is that with them I saw `${Caffe2_MAIN_LIBS}` placed **after** `${Caffe2_DEPENDENCY_LIBS}` so that ld cannot resolve some symbols although `${Caffe2_MAIN_LIBS}` was added before `${Caffe2_DEPENDENCY_LIBS} . With them, I saw messages like:
```
[ 82%] Linking CXX executable ../bin/workspace_test
cd /hack/freedom/ml/caffe2/build_android/caffe2 && /tmp/cmake-3.10.2-Linux-x86_64/bin/cmake -E cmake_link_script CMakeFiles/workspace_test.dir/link.txt --verbose=1
/home/freedom/work/android-ndk-r16/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/bin/arm-linux-androideabi-g++ --sysroot=/home/freedom/work/android-ndk-r16/sysroot  -isystem /home/freedom/work/android-ndk-r16/sysroot/usr/include/arm-linux-androideabi -D__ANDROID_API__=21 -g -DANDROID -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -march=armv7-a -mfloat-abi=softfp -mfpu=vfpv3-d16 -mthumb -Wa,--noexecstack -Wformat -Werror=format-security -frtti -fexceptions -s -Wno-deprecated -std=c++11 -O2 -fPIC -Wno-narrowing -Wno-invalid-partial-specialization -Os -DNDEBUG   -Wl,--exclude-libs,libgcc.a -Wl,--exclude-libs,libatomic.a --sysroot /home/freedom/work/android-ndk-r16/platforms/android-21/arch-arm -Wl,--build-id -Wl,--warn-shared-textrel -Wl,--fatal-warnings -Wl,--fix-cortex-a8 -Wl,--no-undefined -Wl,-z,noexecstack -Wl,-z,relro -Wl,-z,now -Wl,--gc-sections -Wl,-z,nocopyreloc -pie -fPIE  CMakeFiles/workspace_test.dir/core/workspace_test.cc.o  -o ../bin/workspace_test ../lib/libgtest_main.a ../lib/libnnpack.a ../lib/libcpufeatures.a -llog -ldl ../lib/libcpuinfo.a ../lib/libpthreadpool.a -Wl,--whole-archive /hack/freedom/ml/caffe2/build_android/lib/libcaffe2.a -Wl,--no-whole-archive ../lib/libprotobuf.a /home/freedom/work/android-ndk-r16/platforms/android-21/arch-arm/usr/lib/libz.so ../lib/libgtest.a -latomic -lm "/home/freedom/work/android-ndk-r16/sources/cxx-stl/gnu-libstdc++/4.9/libs/armeabi-v7a/libgnustl_static.a" 
/hack/freedom/ml/caffe2/caffe2/share/contrib/nnpack/conv_op.cc:36: error: undefined reference to 'nnp_initialize'
/hack/freedom/ml/caffe2/caffe2/share/contrib/nnpack/conv_op.cc:218: error: undefined reference to 'nnp_convolution_inference'
/hack/freedom/ml/caffe2/caffe2/share/contrib/nnpack/conv_op.cc:254: error: undefined reference to 'nnp_convolution_inference'
/hack/freedom/ml/caffe2/caffe2/share/contrib/nnpack/conv_op.cc:317: error: undefined reference to 'nnp_convolution_inference'
/hack/freedom/ml/caffe2/caffe2/share/contrib/nnpack/conv_op.cc:338: error: undefined reference to 'nnp_convolution_inference'
/hack/freedom/ml/caffe2/caffe2/utils/threadpool/ThreadPool.cc:53: error: undefined reference to 'android_getCpuCount'
/hack/freedom/ml/caffe2/caffe2/utils/threadpool/ThreadPool.cc:70: error: undefined reference to 'android_getCpuIdArm'
collect2: error: ld returned 1 exit status
make[2]: *** [bin/workspace_test] Error 1
make[2]: Leaving directory `/hack/freedom/ml/caffe2/build_android'
make[1]: *** [caffe2/CMakeFiles/workspace_test.dir/all] Error 2
make[1]: Leaving directory `/hack/freedom/ml/caffe2/build_android'
make: *** [all] Error 2
```

With them removed, I can build with ` -DBUILD_TEST=ON -DBUILD_BINARY=ON` successfully.

Maratyszcza replied:
@Yangqing I'm in favor of linking only `${Caffe2_MAIN_LIBS}` and let CMake sort out other dependencies

sf-wind replied:
I get the same error when building caffe2_benchmark. Can we quickly resolve the issues and merge the PR so that the regression can move on? @Yangqing @Maratyszcza 

Thanks.

Yangqing commented on cmake/Utils.cmake:
```diff
@@ -166,7 +166,7 @@ function(caffe2_binary_target target_name_or_src)
   # result, we will explicitly link the test against the Caffe2 dependency
   # libs. This should be changed in a future cleanup.
-  target_link_libraries(${__target} ${Caffe2_DEPENDENCY_LIBS})
+  # target_link_libraries(${__target} ${Caffe2_DEPENDENCY_LIBS})
```
This is still needed at the moment for some tests on server. Could you uncomment it?

freedomtan replied:
ditto

Yangqing reviewed (changes requested):
Thanks!

freedomtan committed:
make -DUSE_OBSERVERS=ON work

```
scripts/build_android.sh -DBUILD_TEST=ON -DBUILD_BINARY=ON  -DBUILD_OBSERVERS=On -DUSE_OBSERVERS=ON
```

freedomtan committed:
add a rule back for non-Android platforms

`-DBUILD_TEST=ON -DBUILD_BINARY=ON -DUSE_OBSERVERS=ON -DBUILD_OBSERVERS=ON`
should work for both Andorid and non-Android platformas (e.g., Ubuntu)

Maratyszcza commented on scripts/build_android.sh:
```diff
@@ -86,6 +86,21 @@ CMAKE_ARGS+=("-DUSE_MOBILE_OPENGL=OFF")
+else
+  CMAKE_ARGS+=("-DCMAKE_EXE_LINKER_FLAGS=-s")
+fi
+
```
Building with `-s` seems reasonable to me, but it should be handled in `CMakeLists.txt` rather than in the build script. We don't expect that all users will use `scripts/build_android.sh`, this script is more of an example.

freedomtan replied:
OK, I'll more them to some cmake files.

freedomtan committed:
move -s to CMakeLists.txt

Maratyszcza commented:
@caffe2bot test this please

Maratyszcza reviewed (approved):
Thank you!

sf-wind commented:
Hi @freedomtan , you said you tested building the observers, but when I tried the merged commit, I get the following errors:
`
[ 98%] Linking CXX executable ../../../../../bin/caffe2_benchmark
/Users/feisun/caffe2/caffe2_origin/caffe2/share/contrib/observers/observer_config.h:42: error: undefined reference to 'caffe2::ObserverConfig::netInitSampleRate_'
/Users/feisun/caffe2/caffe2_origin/caffe2/share/contrib/observers/observer_config.h:42: error: undefined reference to 'caffe2::ObserverConfig::netFollowupSampleRate_'
/Users/feisun/caffe2/caffe2_origin/caffe2/share/contrib/observers/observer_config.h:42: error: undefined reference to 'caffe2::ObserverConfig::netFollowupSampleCount_'
/Users/feisun/caffe2/caffe2_origin/caffe2/share/contrib/observers/observer_config.h:42: error: undefined reference to 'caffe2::ObserverConfig::operatorNetSampleRatio_'
/Users/feisun/caffe2/caffe2_origin/caffe2/share/contrib/observers/observer_config.h:42: error: undefined reference to 'caffe2::ObserverConfig::skipIters_'
collect2: error: ld returned 1 exit status
`

sf-wind commented:
OK, mystery solved. I didn't know there is a new flag called USE_OBSERVERS. I didn't set that one. Thanks for the work!

freedomtan commented:
@sf-wind maybe merge USE_OBSERVERS and BUILD_OBSERVERS?
